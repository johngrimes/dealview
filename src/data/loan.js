export const LoanDefaults = {
  interestRate: 400,
  compoundingPeriod: 'daily',
  lengthInYears: 25,
  repaymentType: 'principalAndInterest',
}

export const loanToLiability = loan => {
  const liability = {}
  liability.type = 'Loan'
  liability.name = loan.name
  liability.startDate = loan.startDate
  liability.endDate = loan.endDate
  if (typeof loan.id === 'string') {
    liability.id = loan.id
  }
  return liability
}
