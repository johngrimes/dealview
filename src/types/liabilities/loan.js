// @flow

import type { Liability } from 'types/liabilities/liability'

export type Loan = {
  +name: string,
  +startDate: string,
  +endDate: string
}

export type LoanWithId = Loan & { +id: string }

export type LoanMap = { +[id: string]: LoanWithId }

export const LoanDefaults = {
  name: '',
  startDate: '',
  endDate: '',
}

export const loanToLiability = (loan: Loan): Liability => {
  const liability = {}
  liability.type = 'Loan'
  liability.name = loan.name
  liability.startDate = loan.startDate
  liability.endDate = loan.endDate
  if (typeof loan.id === 'string') { liability.id = loan.id }
  return liability
}
