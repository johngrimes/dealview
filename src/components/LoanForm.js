import React from 'react'
import PropTypes from 'prop-types'

import InputField from './InputField.js'
import CurrencyField from './CurrencyField.js'
import HiddenField from './HiddenField.js'
import DateField from './DateField.js'
import SelectField from './SelectField.js'
import * as Validations from '../utils/formValidation.js'
import { LoanDefaults } from '../data/loan.js'

import '../styles/forms.css'
import '../styles/buttons.css'

class LoanForm extends React.Component {
  static propTypes = {
    loan: PropTypes.shape({
      name: PropTypes.string,
      startDate: PropTypes.string,
      endDate: PropTypes.string,
      principal: PropTypes.number,
      interestRate: PropTypes.number,
      compoundingPeriod: PropTypes.oneOf([ 'daily', 'monthly' ]),
      lengthInYears: PropTypes.number,
      establishmentFees: PropTypes.number,
      repaymentType: PropTypes.oneOf([ 'principalAndInterest', 'interestOnly' ]),
    }),
    onSubmit: PropTypes.func,
  }

  constructor(props) {
    super(props)
    this.state = {
      loan: LoanDefaults,
      allErrorsShown: false,
      focusedInput: 'name',
    }

    if (props.loan) {
      const loan = props.loan
      this.state = {
        ...this.state,
        loan,
        errors: LoanForm.validate(loan),
        allErrorsShown: true,
      }
    }

    this.handleChange = this.handleChange.bind(this)
    this.handleFocus = this.handleFocus.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  handleChange(fieldName, value): void {
    this.setState(prevState => {
      const loan = { ...prevState.loan, [fieldName]: value }
      return {
        loan,
        errors: LoanForm.validate(loan),
      }
    })
  }

  handleFocus(fieldName): void {
    this.setState(() => ({ focusedInput: fieldName }))
  }

  handleSubmit(event): boolean {
    event.preventDefault()
    this.setState(
      () => ({ errors: LoanForm.validate(this.state.loan) }),
      () => {
        const { errors, loan } = this.state
        if (
          typeof errors !== 'undefined' &&
          Validations.areErrorsPresent(errors)
        ) {
          const firstErrorFieldName = Validations.findFirstErrorFieldName(
            errors
          )
          firstErrorFieldName
            ? this.setState(() => ({
              allErrorsShown: true,
              focusedInput: firstErrorFieldName,
            }))
            : this.setState(() => ({ allErrorsShown: true }))
        } else {
          if (this.props.onSubmit) {
            this.props.onSubmit(loan)
          }
        }
      }
    )
    return false
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.loan) {
      this.setState(() => ({ loan: nextProps.loan }))
    }
  }

  render() {
    const { loan, allErrorsShown, focusedInput } = this.state
    const errors = this.state.errors === undefined ? {} : this.state.errors
    const idField =
      typeof loan.id === 'string'
        ? <HiddenField name='id' value={loan.id} />
        : null

    return (
      <form
        className='loan-form form form-aligned'
        onSubmit={this.handleSubmit}
      >
        <fieldset>
          {idField}
          <InputField
            name='name'
            type='text'
            label='Name'
            value={loan.name}
            errors={errors.name}
            forceErrorDisplay={allErrorsShown}
            onChange={value => this.handleChange('name', value)}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
          <DateField
            name='startDate'
            label='Start date'
            value={loan.startDate}
            errors={errors.startDate}
            forceErrorDisplay={allErrorsShown}
            onChange={value => this.handleChange('startDate', value)}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
          <DateField
            name='endDate'
            label='End date'
            value={loan.endDate}
            errors={errors.endDate}
            forceErrorDisplay={allErrorsShown}
            onChange={value => this.handleChange('endDate', value)}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
          <CurrencyField
            name='principal'
            label='Principal amount'
            value={loan.principal ? loan.principal.toString() : undefined}
            errors={errors.principal}
            forceErrorDisplay={allErrorsShown}
            onChange={value => this.handleChange('principal', value)}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
          <InputField
            name='interestRate'
            type='number'
            min={0}
            label='Interest rate (basis points)'
            value={loan.interestRate ? loan.interestRate.toString() : undefined}
            errors={errors.interestRate}
            forceErrorDisplay={allErrorsShown}
            onChange={value => this.handleChange('interestRate', value)}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
          <SelectField
            name='compoundingPeriod'
            label='Compounding period'
            options={[ [ 'Daily', 'daily' ], [ 'Monthly', 'monthly' ] ]}
            value={loan.compoundingPeriod}
            errors={errors.compoundingPeriod}
            forceErrorDisplay={allErrorsShown}
            onChange={value => this.handleChange('compoundingPeriod', value)}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
          <InputField
            name='lengthInYears'
            type='number'
            min={0}
            label='Length in years'
            value={
              loan.lengthInYears ? loan.lengthInYears.toString() : undefined
            }
            errors={errors.lengthInYears}
            forceErrorDisplay={allErrorsShown}
            onChange={value => this.handleChange('lengthInYears', value)}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
          <CurrencyField
            name='establishmentFees'
            label='Establishment fees'
            value={
              loan.establishmentFees
                ? loan.establishmentFees.toString()
                : undefined
            }
            errors={errors.establishmentFees}
            forceErrorDisplay={allErrorsShown}
            onChange={value => this.handleChange('establishmentFees', value)}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
          <SelectField
            name='repaymentType'
            label='Repayment type'
            options={[ [ 'Daily', 'daily' ], [ 'Monthly', 'monthly' ] ]}
            value={loan.repaymentType}
            errors={errors.repaymentType}
            forceErrorDisplay={allErrorsShown}
            onChange={value => this.handleChange('repaymentType', value)}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
        </fieldset>
        <fieldset>
          <button type='submit' className='button button-primary'>
            Submit
          </button>
        </fieldset>
      </form>
    )
  }

  static validate(loan) {
    const name = []
      .concat(Validations.required(loan.name))
      .concat(Validations.minLength(loan.name, 3))
      .concat(Validations.maxLength(loan.name, 100))

    return { name }
  }
}

export default LoanForm
