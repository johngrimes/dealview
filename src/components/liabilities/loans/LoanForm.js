// @flow

import React from 'react'

import InputField from 'components/forms/InputField'
import HiddenField from 'components/forms/HiddenField'
import DateField from 'components/forms/DateField'
import * as Validations from 'utils/FormValidation'
import { LoanDefaults } from 'types/liabilities/loan'
import { findFirstErrorFieldName } from 'utils/FormValidation'
import type { Loan } from 'types/liabilities/loan'
import type { FieldErrors } from 'utils/FormValidation'

import 'styles/forms.css'
import 'styles/buttons.css'
import 'styles/tables.css'

type LoanErrors = {
  +name: FieldErrors,
  +startDate: FieldErrors,
  +endDate: FieldErrors,
}
const LoanErrorDefaults = {
  name: [],
  startDate: [],
  endDate: [],
}

type Props = {
  +loan?: Loan,
  +onSubmit?: (loan: Loan) => void
}

type State = {
  +loan: Loan,
  +errors?: LoanErrors,
  +allErrorsShown: boolean,
  +focusedInput: string
}

class LoanForm extends React.Component {
  props: Props
  state: State
  handleChange: (fieldName: string, value: string) => void
  handleFocus: (fieldName: string) => void
  handleSubmit: (event: Event) => boolean

  constructor(props: Props) {
    super(props)
    this.state = {
      loan: LoanDefaults,
      allErrorsShown: false,
      focusedInput: 'name',
    }

    if (props.loan) {
      const loan = props.loan
      this.state = {
        ...this.state,
        loan,
        errors: LoanForm.validate(loan),
        allErrorsShown: true,
      }
    }

    this.handleChange = this.handleChange.bind(this)
    this.handleFocus = this.handleFocus.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  handleChange(fieldName: string, value: string): void {
    this.setState(prevState => {
      const loan = { ...prevState.loan, [fieldName]: value }
      return {
        loan,
        errors: LoanForm.validate(loan),
      }
    })
  }

  handleFocus(fieldName: string): void {
    this.setState(() => ({ focusedInput: fieldName }))
  }

  handleSubmit(event: Event): boolean {
    event.preventDefault()
    this.setState(
      () => ({ errors: LoanForm.validate(this.state.loan) }),
      () => {
        const { errors, loan } = this.state
        if (typeof errors !== 'undefined' && Validations.areErrorsPresent(errors)) {
          const firstErrorFieldName = findFirstErrorFieldName(errors)
          firstErrorFieldName
            ? this.setState(() => ({ allErrorsShown: true, focusedInput: firstErrorFieldName }))
            : this.setState(() => ({ allErrorsShown: true }))
        } else {
          if (this.props.onSubmit) { this.props.onSubmit(loan) }
        }
      }
    )
    return false
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.loan) {
      this.setState(() => ({ loan: nextProps.loan }))
    }
  }

  render() {
    const { loan, allErrorsShown, focusedInput } = this.state
    const errors = this.state.errors === undefined ? {} : this.state.errors
    const idField = typeof loan.id === 'string'
      ? <HiddenField name='id' value={loan.id} />
      : null

    return (
      <form className='loan-form form form-aligned' onSubmit={this.handleSubmit}>
        <fieldset>
          {idField}
          <InputField name='name' type='text' label='Name' value={loan.name}
            errors={errors.name} forceErrorDisplay={allErrorsShown}
            onChange={value => this.handleChange('name', value)}
            onFocus={this.handleFocus} focus={focusedInput} />
          <DateField name='startDate' label='Start date'
            value={loan.startDate} errors={errors.startDate}
            forceErrorDisplay={allErrorsShown}
            onChange={value => this.handleChange('startDate', value)}
            onFocus={this.handleFocus} focus={focusedInput} />
          <DateField name='endDate' label='End date'
            value={loan.endDate} errors={errors.endDate}
            forceErrorDisplay={allErrorsShown}
            onChange={value => this.handleChange('endDate', value)}
            onFocus={this.handleFocus} focus={focusedInput} />
        </fieldset>
        <fieldset>
          <button type='submit' className='button button-primary'>Submit</button>
        </fieldset>
      </form>
    )
  }

  static validate(loan: Loan): LoanErrors {
    const name = []
      .concat(Validations.required(loan.name))
      .concat(Validations.minLength(loan.name, 3))
      .concat(Validations.maxLength(loan.name, 100))

    return {
      ...LoanErrorDefaults,
      name,
    }
  }
}

export default LoanForm
