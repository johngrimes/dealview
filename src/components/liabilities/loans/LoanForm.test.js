/* global expect */

import React from 'react'
import { shallow } from 'enzyme'

import LoanForm from 'components/liabilities/loans/LoanForm'
import { validLoan1, validLoan2 } from 'fixtures/loan'

describe('LoanForm', () => {
  it('should render', () => {
    const wrapper = shallow(<LoanForm loan={validLoan1} />)
    expect(wrapper).toMatchSnapshot()
    const wrapper2 = shallow(<LoanForm loan={validLoan2} />)
    expect(wrapper2).toMatchSnapshot()
  })

  it('should focus the first field with errors upon submission', () => {
    const wrapper = shallow(<LoanForm />)
    wrapper.find({ name: 'name' }).prop('onChange')('ab')
    wrapper.find('form').simulate('submit', { preventDefault() {} })
    expect(wrapper.state('focusedInput')).toEqual('name')
  })

  it('should show all errors upon submission', () => {
    const wrapper = shallow(<LoanForm />)
    wrapper.find('form').simulate('submit', { preventDefault() {} })
    expect(wrapper.state('allErrorsShown')).toEqual(true)
  })
})
