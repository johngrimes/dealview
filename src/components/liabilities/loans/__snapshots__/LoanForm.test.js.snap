// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoanForm should render 1`] = `
<form
  className="loan-form form form-aligned"
  onSubmit={[Function]}
>
  <fieldset>
    <InputField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Name"
      name="name"
      onChange={[Function]}
      onFocus={[Function]}
      type="text"
      value="5SS Reno Loan"
    />
    <DateField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Start date"
      name="startDate"
      onChange={[Function]}
      onFocus={[Function]}
      value="2014-09-10"
    />
    <DateField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="End date"
      name="endDate"
      onChange={[Function]}
      onFocus={[Function]}
      value="2039-09-10"
    />
    <InputField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Principal amount"
      min="0"
      name="principal"
      onChange={[Function]}
      onFocus={[Function]}
      type="number"
      value="500000"
    />
    <SelectField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Compounding Period"
      name="compoundingPeriod"
      onChange={[Function]}
      onFocus={[Function]}
      options={
        Array [
          Array [
            "Daily",
            "daily",
          ],
          Array [
            "Monthly",
            "monthly",
          ],
        ]
      }
      value="daily"
    />
    <InputField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Length in years"
      min="0"
      name="lengthInYears"
      onChange={[Function]}
      onFocus={[Function]}
      type="number"
      value="25"
    />
    <InputField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Establishment fees"
      min="0"
      name="establishmentFees"
      onChange={[Function]}
      onFocus={[Function]}
      type="number"
      value="1500"
    />
    <SelectField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Repayment type"
      name="repaymentType"
      onChange={[Function]}
      onFocus={[Function]}
      options={
        Array [
          Array [
            "Daily",
            "daily",
          ],
          Array [
            "Monthly",
            "monthly",
          ],
        ]
      }
      value="principalAndInterest"
    />
  </fieldset>
  <fieldset>
    <button
      className="button button-primary"
      type="submit"
    >
      Submit
    </button>
  </fieldset>
</form>
`;

exports[`LoanForm should render 2`] = `
<form
  className="loan-form form form-aligned"
  onSubmit={[Function]}
>
  <fieldset>
    <InputField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Name"
      name="name"
      onChange={[Function]}
      onFocus={[Function]}
      type="text"
      value="Briggs Primary"
    />
    <DateField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Start date"
      name="startDate"
      onChange={[Function]}
      onFocus={[Function]}
      value="2014-05-10"
    />
    <DateField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="End date"
      name="endDate"
      onChange={[Function]}
      onFocus={[Function]}
      value="2029-05-10"
    />
    <InputField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Principal amount"
      min="0"
      name="principal"
      onChange={[Function]}
      onFocus={[Function]}
      type="number"
      value="150000"
    />
    <SelectField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Compounding Period"
      name="compoundingPeriod"
      onChange={[Function]}
      onFocus={[Function]}
      options={
        Array [
          Array [
            "Daily",
            "daily",
          ],
          Array [
            "Monthly",
            "monthly",
          ],
        ]
      }
      value="monthly"
    />
    <InputField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Length in years"
      min="0"
      name="lengthInYears"
      onChange={[Function]}
      onFocus={[Function]}
      type="number"
      value="30"
    />
    <InputField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Establishment fees"
      min="0"
      name="establishmentFees"
      onChange={[Function]}
      onFocus={[Function]}
      type="number"
    />
    <SelectField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Repayment type"
      name="repaymentType"
      onChange={[Function]}
      onFocus={[Function]}
      options={
        Array [
          Array [
            "Daily",
            "daily",
          ],
          Array [
            "Monthly",
            "monthly",
          ],
        ]
      }
      value="interestOnly"
    />
  </fieldset>
  <fieldset>
    <button
      className="button button-primary"
      type="submit"
    >
      Submit
    </button>
  </fieldset>
</form>
`;
